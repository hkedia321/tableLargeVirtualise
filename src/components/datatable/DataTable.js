import React, { useState } from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { VariableSizeGrid as Grid } from "react-window"
import AutoSizer from "react-virtualized-auto-sizer"
import StyledComponents from "./DataTableStyledComponents"
import ColumnHeading from "./ColumnHeading"

import DataCellGraph from "./DataCellGraph"
import DataCellCampaignId from "./DataCellCampaignId"

const dataTableGraphRecords = [
  {
    heading: "App and Installs",
    key: "aj_app_and_installs",
  },
  {
    heading: "CoH Real Acquisition",
    key: "aj_coh_0w_and_real_acquisition",
  },
]

const TableStyled = styled.table`
  width: 100%;
  table-layout: fixed;
  height: 90vh;
  overflow-y: scroll;
  border-collapse: collapse;
`

const HeaderRowStyled = styled.tr`
  background: #f2f2f2;
  padding: 15px;
`

const DataTable = React.memo(({ data, show }) => {
  const [campaignColumnOpen, setCampaignColumnOpen] = useState(true)
  const [graphColumnsOpenArray, setGraphColumnsOpenArray] = useState(
    new Array(dataTableGraphRecords.length).fill(true)
  )

  const handleToggleGraphColumnShow = (index) => {
    setGraphColumnsOpenArray((prev) =>
      prev
        .slice(0, index)
        .concat(!prev[index])
        .concat(prev.slice(index + 1))
    )
  }
  const handleToggleCampaignIdColumnShow = () => {
    setCampaignColumnOpen((prev) => !prev)
    // for reflowing highchart
    window.dispatchEvent(new Event("resize"))
  }

  // eslint-disable-next-line react/prop-types
  const Cell = ({ columnIndex, rowIndex, style }) => {
    // console.log(rowIndex, columnIndex);
    const rowData = data[rowIndex]
    const isCampaignColumnOpen = campaignColumnOpen

    if (columnIndex === 0) {
      return (
        <DataCellCampaignId
          style={style}
          text={rowData.groups.Campaign.metadata.name}
          isOpen={isCampaignColumnOpen}
        />
      )
    }
    const columnObject = dataTableGraphRecords[columnIndex - 1]
    return (
      <DataCellGraph
        style={style}
        isOpen={graphColumnsOpenArray[columnIndex - 1]}
        columnObject={columnObject}
        rowData={rowData}
      />
    )
  }

  return (
    <>
      {show && (
        <>
          <div className="dataTableWrap">
            <TableStyled data-testid="datatable">
              <thead>
                <HeaderRowStyled>
                  <ColumnHeading
                    handleToggleColumnShow={handleToggleCampaignIdColumnShow}
                    text="Campaign Id"
                    isGraphHeader={false}
                    isOpen={campaignColumnOpen}
                  />
                  {dataTableGraphRecords.map((item, index) => (
                    <ColumnHeading
                      key={index}
                      isGraphHeader
                      text={item.heading}
                      columnKey={item.key}
                      data={data}
                      handleToggleColumnShow={() =>
                        handleToggleGraphColumnShow(index)
                      }
                      isOpen={graphColumnsOpenArray[index]}
                    />
                  ))}
                </HeaderRowStyled>
              </thead>
              <tbody>
                <AutoSizer>
                  {({ height, width }) => (
                    <Grid
                      className="Grid"
                      height={height}
                      columnCount={3}
                      rowCount={data.length}
                      width={width}
                      rowHeight={() => 250}
                      columnWidth={() => 440}
                    >
                      {Cell}
                    </Grid>
                  )}
                </AutoSizer>
              </tbody>
            </TableStyled>
          </div>
        </>
      )}
      <StyledComponents />
    </>
  )
})

DataTable.propTypes = {
  data: PropTypes.array.isRequired,
  show: PropTypes.bool.isRequired,
}
export default DataTable
