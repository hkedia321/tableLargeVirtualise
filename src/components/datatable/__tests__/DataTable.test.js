import React from "react"
import { render, cleanup } from "@testing-library/react"
import "@testing-library/jest-dom"
import renderer from "react-test-renderer"
import "jest-styled-components"
import DataTable from "../DataTable"

jest.mock("react-tooltip", () => () => <div>react-tooltip</div>)

jest.mock(
  "highcharts-react-official",
  () =>
    // eslint-disable-next-line react/prop-types
    function DummyHighcharts({ options }) {
      return <div id="MockedHighcharts">{JSON.stringify(options)}</div>
    }
)

afterEach(cleanup)

const dataTableProps = {
  data: [
    {
      groups: {
        Campaign: {
          metadata: { name: "99733871-f7c5-4b52-9773-1ca69816fc03" },
        },
      },
      trend: [
        {
          aj_app_and_installs: 178,
          aj_app_ios_installs: 0,
          aj_coh_0w_and_real_acquisition: 14,
          aj_coh_0w_ios_real_acquisition: 0,
          day: "2020-06-01",
        },
      ],
    },
  ],
  show: true,
}

it("renders DataTable correctly", () => {
  const { getByTestId, container } = render(<DataTable {...dataTableProps} />)
  expect(getByTestId("datatable"))
  expect(container.querySelectorAll("tr").length).toBeGreaterThan(0)
})

it("renders DataTable correctly when hidden", () => {
  const { queryByTestId } = render(
    <DataTable {...dataTableProps} show={false} />
  )
  expect(queryByTestId("datatable")).toBeNull()
})

it("matches snapshot", () => {
  const tree = renderer.create(<DataTable {...dataTableProps} />).toJSON()
  expect(tree).toMatchSnapshot()
})
