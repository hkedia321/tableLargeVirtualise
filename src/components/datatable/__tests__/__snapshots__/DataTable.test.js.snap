// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
.c3 {
  cursor: pointer;
}

.c2 {
  padding: 10px;
  text-align: right;
  width: auto;
  color: #4b4b4b;
  font-weight: normal;
  padding-bottom: 25px;
}

.c4 {
  padding: 10px;
  text-align: right;
  width: auto;
  color: #4b4b4b;
  font-weight: normal;
  padding-bottom: 10px;
}

.c5 {
  margin: 0px;
  color: #7f7f7f;
  font-weight: normal;
}

.c0 {
  width: 100%;
  table-layout: fixed;
  height: 90vh;
  overflow-y: scroll;
  border-collapse: collapse;
}

.c1 {
  background: #f2f2f2;
  padding: 15px;
}

<div
  className="dataTableWrap"
>
  <table
    className="c0"
    data-testid="datatable"
  >
    <thead>
      <tr
        className="c1"
      >
        <th
          className="c2"
        >
          Campaign Id
           
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-caret-down fa-w-10 c3"
            data-for="columnCollapseTooltip"
            data-icon="caret-down"
            data-prefix="fas"
            data-tip={true}
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <div>
            react-tooltip
          </div>
        </th>
        <th
          className="c4"
        >
          App and Installs
           
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-caret-down fa-w-10 c3"
            data-for="columnCollapseTooltip"
            data-icon="caret-down"
            data-prefix="fas"
            data-tip={true}
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <h4
            className="c5"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&euro;&nbsp;178",
              }
            }
          />
          <div>
            react-tooltip
          </div>
        </th>
        <th
          className="c4"
        >
          CoH Real Acquisition
           
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-caret-down fa-w-10 c3"
            data-for="columnCollapseTooltip"
            data-icon="caret-down"
            data-prefix="fas"
            data-tip={true}
            focusable="false"
            onClick={[Function]}
            role="img"
            style={Object {}}
            viewBox="0 0 320 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          <h4
            className="c5"
            dangerouslySetInnerHTML={
              Object {
                "__html": "&euro;&nbsp;14",
              }
            }
          />
          <div>
            react-tooltip
          </div>
        </th>
      </tr>
    </thead>
    <tbody>
      <div
        style={
          Object {
            "height": 0,
            "overflow": "visible",
            "width": 0,
          }
        }
      />
    </tbody>
  </table>
</div>
`;
